spec/unit/rust-crypto/verification.spec.ts

```diff
@@ -152,3 +405,148 @@ function makeMockedInner(): Mocked<RustSdkCryptoJs.VerificationRequest> {
+    user_signing_key: any;
+}
+
+/** create an Olm machine and device/cross-signing keys for a user */
+async function initOlmMachineAndKeys(
+    userId: string,
+    deviceId: string,
+): Promise<[RustSdkCryptoJs.OlmMachine, IDeviceKeys, CrossSigningKeys]> {
+    const olmMachine = await RustSdkCryptoJs.OlmMachine.initialize(
+        new RustSdkCryptoJs.UserId(userId),
+        new RustSdkCryptoJs.DeviceId(deviceId),
+        undefined,
+        undefined,
+    );
+    const { uploadKeysRequest, uploadSignaturesRequest, uploadSigningKeysRequest } =
+        await olmMachine.bootstrapCrossSigning(true);
+    const deviceKeys = JSON.parse(uploadKeysRequest.body).device_keys;
+    await olmMachine.markRequestAsSent(
+        uploadKeysRequest.id,
+        uploadKeysRequest.type,
+        '{"one_time_key_counts":{"signed_curve25519":100}}',
+    );
+    const crossSigningSignatures = JSON.parse(uploadSignaturesRequest.body);
+    if (crossSigningSignatures[userId] && crossSigningSignatures[userId][deviceId]) {
+        for (const [keyId, signature] of Object.entries(
+            crossSigningSignatures[userId][deviceId]["signatures"][userId],
+        )) {
+            deviceKeys["signatures"][userId][keyId] = signature;
+        }
+    }
+    const crossSigningKeys = JSON.parse(uploadSigningKeysRequest.body);
+    // note: the upload signatures request and upload signing keys requests
+    // don't need to me marked as sent in the Olm machine
+
+    return [olmMachine, deviceKeys, crossSigningKeys];
+}
+
+/** Loop for handling outgoing requests from an Olm machine.
+ *
+ * Simulates a server with two users: "us" and "them".  Handles key query
+ * requests, querying either our keys or the other user's keys.  Room messages
+ * are sent as incoming verification events to the other user.  A custom
+ * handler can be added to override default request processing (the handler
+ * should return a response body to inhibit default processing).
+ *
+ * Can also be used as an OutgoingRequestProcessor. */
+function makeRequestLoop(
+    ourOlmMachine: RustSdkCryptoJs.OlmMachine,
+    ourDeviceKeys: IDeviceKeys,
+    ourCrossSigningKeys: CrossSigningKeys,
+    theirOlmMachine: RustSdkCryptoJs.OlmMachine,
+    theirDeviceKeys: IDeviceKeys,
+    theirCrossSigningKeys: CrossSigningKeys,
+    customHandler:
+        | ((request: OutgoingRequest | RustSdkCryptoJs.UploadSigningKeysRequest) => Promise<any>)
+        | undefined = undefined,
+) {
+    let stopRequestLoop = false;
+    const ourUserId = ourOlmMachine.userId.toString();
+    const ourDeviceId = ourOlmMachine.deviceId.toString();
+    const theirUserId = theirOlmMachine.userId.toString();
+    const theirDeviceId = theirOlmMachine.deviceId.toString();
+
+    async function makeOutgoingRequest(
+        request: OutgoingRequest | RustSdkCryptoJs.UploadSigningKeysRequest,
+    ): Promise<any> {
+        let resp: any = customHandler && (await customHandler(request));
+        if (resp == undefined) {
```

Comment:
```suggestion
        if (resp === undefined) {
```

Reply:
It would be quite nice to stick this whole block into a separate `handleRequest(ourOlmMachine, theirOlmMachine)` function.